### PR TITLE
feat(#202): CachingTransport library API with stale-fallback policy

### LINKED ISSUE #164: proxy: cache miss + upstream down — response behavior for BOTH non-SSE and SSE endpoints
PLACEHOLDER

### LINKED ISSUE #202: feat: CachingTransport library API — http.RoundTripper that replays from store on hit, records + forwards on miss
## Problem statement

httptape's existing library surface exposes `Recorder` (forwards + records, no replay) and `Server` (replays only, no forward). The `proxy` CLI subcommand combines both — cache by match, fall through on miss — but that logic lives inside the CLI binary, not as a library primitive. Consumers that embed httptape as a Go library need the combined logic exposed as an `http.RoundTripper`.

## User stories

**As a** Go developer embedding httptape in a Caddy-based egress proxy,
**I want** an `http.RoundTripper` that replays from a tape store on match and records + forwards on miss,
**so that** I can drop it into any Go HTTP client or middleware chain without running a sidecar process.

**As a** demo platform operator,
**I want** to record ~50 real LLM responses once and replay them infinitely for every demo visitor at zero API cost,
**so that** demo hosting costs ~EUR 5/month on a single VM instead of burning API credits on every pageview.

## Context

VibeWarden will embed httptape as a Go library inside its Caddy-based egress proxy. Single-binary deployment is VibeWarden's core positioning — a sidecar httptape container is not acceptable. Two concrete use cases drive priority:

**A. Zero-cost demo hosting (LinkGuard)**: Record real LLM responses, replay infinitely. VibeWarden middleware inspects real recorded responses (not fabricated), so prompt-injection detection fires genuinely. httptape sanitizes API keys on record. Fixtures committed to the repo — no runtime API dependency.

**B. Egress proxy integration**: Record egress traffic in dev/staging for replay in tests. Integration tests against recorded fixtures instead of live APIs. Upstream API drift detection.

Both need the same primitive: an `http.RoundTripper` that replays from a store on match, records on miss.

## Proposed API surface

```go
// NewCachingTransport wraps an upstream transport with store-backed caching.
// On RoundTrip, it consults the store for a matching tape. On hit, it returns
// the recorded response. On miss, it forwards to upstream, records the response
// via the sanitization pipeline (if configured), and returns it.
func NewCachingTransport(
    upstream http.RoundTripper,
    store Store,
    opts ...CachingOption,
) *CachingTransport

type CachingOption func(*CachingTransport)

// WithCacheMatcher sets the matcher used to identify equivalent requests.
// Defaults to method + path + body_hash (the right default for LLM caching:
// same prompt body = same cached response).
func WithCacheMatcher(m Matcher) CachingOption

// WithCacheSanitizer wraps recorded tapes with the given pipeline before
// they're persisted. Same semantics as Recorder.
func WithCacheSanitizer(p *Pipeline) CachingOption

// WithCacheFilter skips caching for responses that don't match the filter.
// Default: cache 2xx responses only. Users can broaden to include 3xx/4xx
// or narrow further (e.g., cache only JSON responses).
func WithCacheFilter(fn func(*http.Response) bool) CachingOption

// WithSingleFlight deduplicates concurrent misses for the same match key —
// only one upstream call, others wait for the result. Defaults to true.
// Disable only if you want "accidentally record multiple variants of the
// same request" behavior (rare).
func WithSingleFlight(enabled bool) CachingOption
```

## RoundTrip flow

1. Read and buffer the request body (for hash matching AND for the upstream call).
2. Compute the match candidate via the configured Matcher.
3. Look up in the store.
4. **Hit**: return the recorded response (headers, body, SSE events verbatim).
5. **Miss**:
   a. If single-flight, coordinate via per-key mutex so concurrent misses collapse to one upstream call.
   b. Forward to upstream.
   c. If response matches the cache filter, persist via the sanitization pipeline.
   d. Return the response.
6. **Upstream fails on miss**: propagate the error. If there is a partial/stale match available AND #164's fallback policy is in effect, serve the stale one with a diagnostic header (architect decides exact interaction with #164).

## Streaming / SSE

Tee semantics: when forwarding an SSE upstream response, the client receives each event AS it arrives AND the store accumulates the complete event list. On cache hit, replay via the existing `serveSSE` path. Architect to confirm the teeing story works cleanly with sanitization on SSE events.

## Acceptance criteria

- [ ] `CachingTransport` struct implementing `http.RoundTripper`.
- [ ] `NewCachingTransport` constructor with the options listed above.
- [ ] Default matcher is `NewCompositeMatcher(MethodCriterion{}, PathCriterion{}, BodyHashCriterion{})` — appropriate for LLM caching.
- [ ] Default cache filter: status 2xx only.
- [ ] Default single-flight: enabled.
- [ ] Thread-safe — passes `go test -race ./...`.
- [ ] Body handling correctness: request body is buffered AND restored so the upstream call sees it intact.
- [ ] SSE support: tee semantics on miss, replay from store on hit.
- [ ] Sanitization pipeline integration: the existing `Pipeline` runs on the miss-path before store write.
- [ ] Interaction with #164 edge cases documented: what happens when cache miss AND upstream down? If stale-match policy is enabled (future), serve stale; otherwise propagate error.
- [ ] Tests: unit (hit, miss, upstream error, filter rejects response, sanitization applies, single-flight dedup), integration (end-to-end with real HTTP client + `httptest` upstream + MemoryStore).
- [ ] Docs: new `docs/caching-transport.md`, update `docs/proxy.md` to cross-reference the library primitive vs the CLI subcommand, update `docs/api-reference.md`.
- [ ] Example in docs showing the LLM-caching pattern:
  ```go
  transport := httptape.NewCachingTransport(
      http.DefaultTransport,
      store,
      httptape.WithCacheSanitizer(pipeline),
  )
  client := &http.Client{Transport: transport}
  // Use client exactly like http.DefaultClient — caching is transparent.
  ```
- [ ] CHANGELOG entry under v0.13.0.

## Open questions for the architect

1. **Naming**: `CachingTransport` (focus on behavior) vs `ReplayTransport` (focus on fixture-replay semantics) vs `TapedTransport` (brand-consistent). PM recommendation: `CachingTransport` reads naturally for the LLM-cost use case; "replay" is slightly misleading since it also records.

2. **Error handling on store failure during record**: if the upstream call succeeds but store persistence fails, return the response anyway (log the failure) or propagate the error? PM recommends return-and-log — cache failures should not break the user's application.

3. **How does CachingTransport relate to the existing proxy CLI subcommand**? Architect should decide whether the CLI subcommand is refactored to use CachingTransport internally (unifying the code paths) or left as-is (two implementations). PM lean: refactor the CLI to use CachingTransport — single source of truth.

4. **Cache-control header respect**: does `CachingTransport` honor upstream `Cache-Control: no-store`? PM recommendation: NO — this is not an RFC 7234 cache; it is a tape-match layer. LLM APIs return `no-store` and that is exactly the use case we need to override. Document prominently.

5. **Body size limits for hashing**: request body gets fully buffered for hash computation. Memory risk on large uploads. PM recommends a configurable `WithMaxBodySize(n int)` option (default e.g. 10 MiB); requests with body > limit bypass cache and forward directly.

6. **TTL / eviction**: fixtures are conceptually immutable (a tape archive grows forever). Should `CachingTransport` support TTL for demo hygiene (auto-expire old recordings)? PM recommends NO for v1 — keep it simple. Users can prune manually if needed.

7. **Interaction with existing Recorder**: `Recorder` today is "forwards + records" (no replay). Should it be deprecated in favor of CachingTransport? PM recommends KEEPING Recorder as a record-only primitive — useful when you want to capture every call regardless of matches (e.g., audit use cases). CachingTransport is the "replay + record-on-miss" sibling.

## Dependencies

- No hard dependencies. Can ship in v0.13 standalone.
- Soft dependency on #164 resolution for the "cache miss + upstream down" edge case policy (architect should review #164's open questions and fold the answers into CachingTransport's behavior).
- No dependency on #194 (journal), #195 (scenarios), #196 (templates), #197 (TLS), #199 (synthesis).

## Out of scope

- No admin endpoints for cache control (flush, stats) — file as follow-up if demand surfaces.
- No custom serialization formats — uses the existing Tape JSON format.
- No HTTP/2 or gRPC specifics — `http.RoundTripper` handles HTTP/1.1 and HTTP/2 transparently via stdlib.
- No "stale-while-revalidate" semantics — #164 covers the stale-on-upstream-down story; this issue respects that but does not define it.
- No TTL or eviction — fixtures are conceptually immutable.
- No deprecation of existing Recorder.

### PR DESCRIPTION
**Pull request by tibtof**
Closes #202, closes #164

## Summary

- Add `CachingTransport`, a new `http.RoundTripper` that provides transparent, single-store cache-through-upstream behavior. On cache hit, returns recorded response without contacting upstream. On cache miss, forwards to upstream, records (with optional sanitization), and returns.
- Implements all 10 ADR-44 decisions: naming (CachingTransport), store-failure-on-record (return-and-log), no Cache-Control respect, configurable body size limits, no TTL/eviction, Recorder not deprecated, stale-fallback policy, SSE partial disconnect detection, upstream timeout, single-flight dedup.
- 10 functional options: `WithCacheMatcher`, `WithCacheSanitizer`, `WithCacheFilter`, `WithCacheSingleFlight`, `WithCacheMaxBodySize`, `WithCacheRoute`, `WithCacheOnError`, `WithCacheSSERecording`, `WithCacheUpstreamDownFallback`, `WithCacheUpstreamTimeout`.
- Single-flight deduplication (stdlib-only, `sync.Mutex` + `map[string]*sfCall`) for concurrent identical cache misses.
- Stale-fallback policy (#164): opt-in via `WithCacheUpstreamDownFallback(true)`, returns cached tape with `X-Httptape-Stale: true` header when upstream is unreachable.
- SSE tee recording with `eofTrackingReadCloser` for partial disconnect detection.
- 20+ test cases covering all ADR-44 test strategy items, including race condition tests.
- Documentation: new `docs/caching-transport.md`, updates to `docs/proxy.md`, `docs/api-reference.md`, `docs/recording.md`, `docs/cli.md`, `llms.txt`, `llms-full.txt`, `CHANGELOG.md`.

## Test plan

- [x] `go build ./...` passes
- [x] `go test ./...` passes (20+ new tests)
- [x] `go vet ./...` passes
- [x] `go test -race ./...` passes
- [x] All existing tests continue to pass (no regressions)
- [x] Cache hit (non-SSE and SSE)
- [x] Cache miss with recording (non-SSE and SSE tee)
- [x] SSE partial disconnect detection (tape discarded)
- [x] Upstream error propagation
- [x] Stale fallback (hit and miss paths)
- [x] Cache filter rejection (non-2xx not cached)
- [x] Sanitization applied to stored tapes
- [x] Single-flight dedup (upstream called exactly once for concurrent identical requests)
- [x] MaxBodySize bypass (oversize bodies forwarded without cache interaction)
- [x] Store.Save failure handling (response still returned, onError called)
- [x] Upstream timeout with stale fallback
- [x] Route filtering
- [x] Concurrent safety (race detector clean)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: caching_transport.go [hunk 1]
```diff
@@ -0,0 +1,687 @@
...
+					ct.onErrorSafe(fbErr)
+				}
+				if resp != nil {
+					return resp, nil
+				}
+			}
+			return nil, call.err
+		}
+		// Clone the response for each waiter.
+		return cloneResponse(call.resp, call.body), nil
```
Comment (tibtof) [inline caching_transport.go:325]: **Correctness bug: SSE single-flight waiters receive an empty response body.**

When the leader returns an SSE response (line 330-339), `call.resp` is set but `call.body` is never set (remains nil). When waiters wake up (line 305-306), they call `cloneResponse(call.resp, call.body)` which creates a response with `bytes.NewReader(nil)` -- an empty body.

ADR-44 specifies: *"Concurrent callers with the same key WAIT for the first caller's stream to complete, then get the cached tape (replay)."* But the current code does not re-query the store after the leader completes. Waiters silently receive a hollow response.

**Impact**: any concurrent SSE requests with single-flight enabled will produce empty responses for all callers except the leader.

**Fix**: After `call.wg.Wait()`, if the leader's response was SSE (detectable via Content-Type header on `call.resp`), the waiter should re-query the store via `ct.store.List` + `ct.matcher.Match` and serve the cached SSE tape via `ct.sseResponseFromTape`. Fall back to `cloneResponse` only for non-SSE responses. Add a `TestCachingTransport_SingleFlightSSEWaiters` test that verifies concurrent SSE requests produce correct event bodies for all callers.

File: docs/proxy.md [hunk 2]
```diff
@@ -265,8 +265,15 @@ proxy := httptape.NewProxy(l1, l2, httptape.WithProxySanitizer(sanitizer))
 
 `Proxy` is safe for concurrent use by multiple goroutines. `RoundTrip` may be called from multiple goroutines simultaneously.
 
+## CachingTransport
+
+The `Proxy` type internally composes a `CachingTransport` for the L2+upstream path. If you need a simpler, single-store caching `http.RoundTripper` without the L1/L2 split, use `CachingTransport` directly. It supports single-flight deduplication, stale-fallback, SSE tee recording, and sanitize-on-write -- the same core features as Proxy but in a lighter package.
```
Comment (tibtof) [inline docs/proxy.md]: **Documentation accuracy: false claim that Proxy composes CachingTransport internally.**

This statement is factually incorrect because the proxy refactor was deferred. The Proxy type does NOT currently compose CachingTransport -- it still uses its own independent cache-through-upstream implementation. This same false claim appears in 7 locations across the PR:

1. `docs/proxy.md` line 270
2. `docs/caching-transport.md` line 12 and line 304
3. `docs/cli.md` line 81
4. `llms.txt` line 26 and line 122
5. `llms-full.txt` line 26 and line 138

Shipping documentation that claims composition exists when it does not will confuse both human readers and LLM-based tools consuming llms.txt. Since the proxy refactor is deferred, these should be changed to future-tense or aspirational language, e.g.: "The Proxy type can be refactored to compose CachingTransport internally (planned, see ADR-44)." or simply removed until the refactor lands.

**Fix**: update all 7 locations to accurately reflect the current state. Use language like "CachingTransport provides the same core logic that Proxy uses for L2+upstream, but as a standalone single-store primitive" instead of claiming composition that does not exist.

File: caching_transport.go [hunk 3]
```diff
@@ -0,0 +1,687 @@
...
+
+// roundTripWithSingleFlight coordinates single-flight dedup for non-SSE
+// cache misses. SSE responses are handled differently: the first caller
+// gets the live tee'd stream and concurrent callers wait for completion,
+// then get the cached tape.
+func (ct *CachingTransport) roundTripWithSingleFlight(req *http.Request, reqBody []byte, bodyHash, sfKey string) (*http.Response, error) {
+	ct.sfMu.Lock()
+	if call, ok := ct.sfCalls[sfKey]; ok {
+		ct.sfMu.Unlock()
+		call.wg.Wait()
```
Comment (tibtof) [inline caching_transport.go:299]: **Concurrency concern: single-flight waiters do not respect context cancellation.**

When a waiter enters line 288 (`call.wg.Wait()`), it blocks indefinitely until the leader completes the upstream call. If the waiter's request context is cancelled (client disconnect, overall timeout), the goroutine remains blocked -- there is no select on `req.Context().Done()`.

This means a cancelled HTTP client will leak a blocked goroutine until the leader's upstream call finishes. For long-running SSE streams, this could be minutes.

**Impact**: goroutine leak under cancellation. Functional correctness is preserved (the waiter will eventually return), but resource usage is unbounded if clients frequently cancel.

**This is not a blocking issue** for this PR, since `sync.WaitGroup` does not support context-aware waiting in stdlib. However, it should be documented as a known limitation in the godoc for `WithCacheSingleFlight`, and a simple mitigation (channel-based wait with select) could be added in a follow-up. A one-line godoc addition like "Note: waiters blocked on a shared upstream call do not exit early if their request context is cancelled" would suffice for now.

**Comment by tibtof** [comment]
## Reviewer Verdict: CHANGES REQUESTED

(Note: posted as comment because git user matches PR author -- cannot use `gh pr review --request-changes`.)

### Blocking Issues (2)

**1. SSE single-flight waiter correctness bug** (`caching_transport.go:306`)
When single-flight is enabled and the leader returns an SSE response, `call.body` is never set (remains nil). Waiters call `cloneResponse(call.resp, call.body)` which produces a response with an empty body. ADR-44 specifies waiters should re-query the store after the leader's stream completes and serve the cached SSE tape. The current code does not do this, and there is no test covering SSE single-flight waiters. Fix the waiter path to re-query the store for SSE responses, and add a `TestCachingTransport_SingleFlightSSEWaiters` test.

**2. Documentation falsely claims Proxy composes CachingTransport** (7 locations)
Since the proxy refactor was deferred, 7 locations across `docs/proxy.md`, `docs/caching-transport.md`, `docs/cli.md`, `llms.txt`, and `llms-full.txt` incorrectly state that Proxy composes CachingTransport internally. This is factually false -- Proxy still uses its own independent implementation. These must be corrected to reflect the current state before merge.

### Non-blocking Concern (1)

**3. Context cancellation in single-flight waiters** (`caching_transport.go:288`)
`call.wg.Wait()` blocks indefinitely regardless of the waiter's request context. Acceptable for v0.13, but should be documented as a known limitation in the `WithCacheSingleFlight` godoc.

### Proxy Refactor Deferral -- Accepted with Condition

I accept the proxy refactor deferral. The library primitive delivers the primary value and unblocks VibeWarden's integration. However, **approval is conditional on a follow-up issue being filed** for the proxy refactor per ADR-44 Option B. The orchestrator should file this before final approval.

### What Passed Review

- All 10 ADR-44 decisions correctly reflected in code (1 implementation bug noted above)
- Non-SSE single-flight, body buffering, SSE tee, stale fallback, sanitization all correct
- 27 tests, all stdlib-only, good coverage
- Architecture, code quality, Go idioms, dependencies all clean
- No backward compatibility regressions
- Documentation structure and cross-references complete (content accuracy issue noted above)

See inline comments for details on each issue.

File: caching_transport.go (hunk 1 above)
Comment (tibtof) [reply caching_transport.go:325]: Fixed in 6edbe80. The `sfCall` struct now tracks an `sse` flag. When the leader's response is SSE, the flag is set before `wg.Done()`. In the waiter path, if `call.sse` is true, waiters re-query the store via the new `reQueryStoreForSSE` method instead of calling `cloneResponse`. If the tape is not found (leader's store write failed or stream was truncated), the waiter falls through to a direct upstream call. Added `TestCachingTransport_SingleFlightSSEWaiters` covering 5 concurrent SSE requests with a barrier to ensure contention.

File: docs/proxy.md (hunk 2 above)
Comment (tibtof) [reply docs/proxy.md]: Fixed in 6edbe80. Corrected all 7 locations across 5 files:

1. `docs/proxy.md` line 270 -- now states Proxy and CachingTransport are separate implementations, unification planned (#205)
2. `docs/caching-transport.md` line 12 -- table row updated to reference separate implementation + #205
3. `docs/caching-transport.md` line 304 -- comparison section updated to accurately describe separate implementations
4. `docs/cli.md` line 81 -- proxy subcommand description corrected
5. `llms.txt` line 26 -- Core Types list corrected
6. `llms.txt` line 122 -- Proxy section corrected
7. `llms-full.txt` line 26, 138, 1896 -- all three occurrences corrected

All replacements use accurate language: "separate implementations" + "unification planned, see #205".

File: caching_transport.go (hunk 3 above)
Comment (tibtof) [reply caching_transport.go:299]: Fixed in 6edbe80. Added a known-limitation note to the `WithCacheSingleFlight` godoc documenting that waiters do not observe context cancellation until the leader's upstream call completes. The godoc explains: (a) the limitation exists, (b) why (`sync.WaitGroup` is not context-aware), (c) workarounds (disable single-flight or use `WithCacheUpstreamTimeout`). Also added a "Known limitations" section to `docs/caching-transport.md` with the same information.

**Comment by tibtof** [comment]
## Reviewer Verdict: APPROVED

(Posted as comment because git user matches PR author.)

All three findings from the previous review cycle are correctly addressed in commit 6edbe80.

### Fix 1 -- SSE single-flight waiter bug (VERIFIED)

- `sfCall.sse` field added; leader sets it before `wg.Done()`
- `reQueryStoreForSSE` method correctly handles: (a) normal case (re-queries store, serves cached SSE tape), (b) store failure (logs via `onErrorSafe`, falls back to direct upstream), (c) no matching tape (logs, falls back to direct upstream)
- `TestCachingTransport_SingleFlightSSEWaiters` exercises 5 concurrent goroutines with barrier sync, verifies all receive all 3 SSE events
- Race-safe: `go test -race ./...` passes

### Fix 2 -- Doc corrections (VERIFIED)

All locations across `docs/proxy.md`, `docs/caching-transport.md`, `docs/cli.md`, `llms.txt`, and `llms-full.txt` now accurately describe CachingTransport and Proxy as "separate implementations, unification planned in #205." No false claims remain.

### Fix 3 -- Context-cancellation limitation (VERIFIED)

- `WithCacheSingleFlight` godoc (`caching_transport.go:148-157`) clearly explains the `sync.WaitGroup`-based limitation and two workarounds
- `docs/caching-transport.md` Known Limitations section provides the same guidance in user-facing docs

### Approval condition -- #205 (CONFIRMED)

Issue #205 is filed in the `v0.13` milestone with `type:refactor` + `priority:medium` labels. Status: `READY_FOR_ARCH`. The deferral will not become permanent forgetting.

### Additional checks

- All 10 ADR-44 decisions correctly reflected in code
- `go test -race ./...` green (7.7s)
- Architecture, dependencies, code quality all clean
- 29 tests, all stdlib-only
- No external dependencies added

Ready for human review and merge.
